src/adapter/src/coord/timeline.rs

```diff
@@ -481,65 +502,128 @@ impl Coordinator {
         global_timelines.get_mut(timeline).expect("inserted above")
     }
 
-    pub(crate) fn remove_storage_ids_from_timeline<I>(&mut self, ids: I) -> Vec<Timeline>
+    pub(crate) fn associate_with_timelines(
+        &self,
+        storage_ids: impl IntoIterator<Item = GlobalId> + Clone,
+        compute_ids: impl IntoIterator<Item = (ComputeInstanceId, GlobalId)> + Clone,
+        clusters: impl IntoIterator<Item = ComputeInstanceId> + Clone,
+    ) -> BTreeMap<Timeline, (bool, Vec<TimelineAssociation>)> {
+        // Map our resources to Timelines.
+        let mut timelines: BTreeMap<_, (bool, Vec<TimelineAssociation>)> = BTreeMap::new();
+
+        for id in storage_ids.into_iter() {
+            if let TimelineContext::TimelineDependent(timeline) = self.get_timeline_context(id) {
+                let assoc = TimelineAssociation::StorageId(id);
+                if let Some(assocs) = timelines.get_mut(&timeline) {
+                    assocs.1.push(assoc);
+                } else {
+                    timelines.insert(timeline, (false, vec![assoc]));
+                }
+            }
+        }
+
+        for (compute_id, id) in compute_ids.into_iter() {
+            if let TimelineContext::TimelineDependent(timeline) = self.get_timeline_context(id) {
+                let assoc = TimelineAssociation::ComputeId((compute_id, id));
+                if let Some(assocs) = timelines.get_mut(&timeline) {
+                    assocs.1.push(assoc);
+                } else {
+                    timelines.insert(timeline, (false, vec![assoc]));
+                }
+            }
+        }
+
+        let clusters: BTreeSet<_> = clusters.into_iter().collect();
+        for (timeline, TimelineState { read_holds, .. }) in &self.global_timelines {
+            for id in read_holds
+                .compute_ids()
+                .map(|(cluster_id, _id)| *cluster_id)
+                .filter(|cluster_id| clusters.contains(cluster_id))
+            {
+                let assoc = TimelineAssociation::ComputeInstance(id);
+                if let Some(assocs) = timelines.get_mut(timeline) {
+                    assocs.1.push(assoc);
+                } else {
+                    timelines.insert(timeline.clone(), (false, vec![assoc]));
+                }
+            }
+        }
+
+        // Check if any of our Timelines would be empty, after removing all of
+        // the associated resources.
+        for (timeline, (empty, resources)) in &mut timelines {
+            let TimelineState { read_holds, .. } = self
+                .global_timelines
+                .get(timeline)
+                .expect("all timeslines have a timestamp oracle");
+
+            // Clone the read holds which we'll use to check for emptiness.
+            //
+            // Note: there might be a more memory efficient, but complex, way
+            // to do this check, for now we're opting to keep the logic simple.
```

Comment:
If resources here was a `CollectionIdBundle` then we could probably do something like 
```Rust
resources.difference(read_holds.id_bundle()).is_empty();
```

Reply:
beautiful, updated the PR to do this!